src/grammar-builder/grammar.ts

```diff
@@ -252,26 +258,48 @@ export function allowedSyntaxKinds (
     }
 }
 
-export function preParse (
-    preParse : Rule,
-    postParse : Rule
-) : Rule {
-    if (typeof preParse == "string") {
-        preParse = seqNoFlatten(preParse);
+export function fieldLengthCheck (
+    field : string,
+    minLength : number,
+    maxLength : number,
+    rule : Rule
+) : RuleObj {
+    if (typeof rule == "string") {
+        rule = seqNoFlatten(rule);
     }
 
+    const fieldCheckArr = rule.fieldCheckArr ?? [];
     return {
-        ...preParse,
-        postParse,
+        ...rule,
+        fieldCheckArr : [
+            ...fieldCheckArr,
+            {
+                type : "FieldLengthCheck",
+                field,
+                minLength,
+                maxLength,
+            },
+        ],
     };
 }
 
+export interface FieldLengthCheck {
+    type : "FieldLengthCheck",
+    field : string,
+    minLength : number,
+    maxLength : number,
+}
+
+export type FieldCheck =
+    | FieldLengthCheck
+;
+
 export interface RuleBase {
     precedence? : number;
     customExtraName? : string;
     penalizeErrorStart? : boolean;
     allowedSyntaxKinds? : string[];
-    postParse? : Rule,
+    fieldCheckArr? : FieldCheck[];
```

Comment:
For new checks.

We can now have a parse tree that is syntactically valid, but semantically invalid because some fields fail some checks.

I wanted to avoid adding semantic checks to the parser, but the `FieldLengthCheck` was so useful and neatly solved error reporting for some invalid syntaxes that I felt compelled to include it.

Reply:
For example,
```sql
SELECT DATE(a, b, c, d, e, f)
```

The `DATE` function only takes 1 argument.

Ideally, I wanted the errors to be,
+ Unexpected ,
+ Unexpected b
+ Unexpected ,
+ Unexpected c
+ Unexpected ,
+ Unexpected d
+ Unexpected ,
+ Unexpected e
+ Unexpected ,
+ Unexpected f

-----

However, the parser tries to **minimize** the number of errors. So, we maybe get something like,
+ Unexpected DATE
+ Unexpected (
+ Unexpected )

And the parser just sees,
```
SELECT a, b, c, d, e, f
```

Because 3 errors is fewer than 10 errors.

Reply:
Part of the problem is that some functions are "special" in MySQL and have specific rules declared for them, rather than using a "generic" function call rule.

https://github.com/mysql/mysql-server/blob/5c8c085ba96d30d697d0baa54d67b102c232116b/sql/sql_yacc.yy#L9623

Reply:
And I want the parser here to give a syntax error whenever we have a MySQL syntax error, as much as possible.

So, I added field checks.

We parse `DATE(a, b, c, d, e, f)` in its entirety.
This gives us,
```js
item = [Identifier, Identifier, Identifier, Identifier, Identifier, Identifier]
```

The field check then checks if `item.length == 1`.
If it isn't, report an error.

Reply:
This also fits more naturally with how a human would parse the above SQL.

A human would look at `DATE(a, b, c, d, e, f)` and say, "Oh, you just have too many arguments."

They wouldn't say, "Unexpected DATE, (, )" or "Unexpected b, c, d, e, f"